plugin/pulpcore/plugin/stages/content_unit_stages.py

```diff
@@ -0,0 +1,176 @@
+                    for field in result.natural_key_fields():
+                        if getattr(declarative_content.content, field) != getattr(result, field):
+                            not_same_unit = True
+                            break
+                    if not_same_unit:
+                        continue
+                    declarative_content.content = result
+        for declarative_content in declarative_content_list:
+            await out_q.put(declarative_content)
+        declarative_content_list = []
+        if shutdown:
+            break
+    await out_q.put(None)
+
+
+async def content_unit_saver(in_q, out_q):
+    """
+    Stages API stage that saves :attr:`DeclarativeContent.content` objects and saves its related
+    :class:`~pulpcore.plugin.models.ContentArtifact` and
+    :class:`~pulpcore.plugin.models.RemoteArtifact` objects too.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    Each "unsaved" Content objects is saved and a :class:`~pulpcore.plugin.models.ContentArtifact`
+    and :class:`~pulpcore.plugin.models.RemoteArtifact` objects too. This allows Pulp to refetch the
+    Artifact in the future if the local copy is removed.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to after it has been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): Each item is a
+            :class:`~pulpcore.plugin.stages.DeclarativeContent`
+        out_q (:class:`asyncio.Queue`): Each item is a
+            :class:`~pulpcore.plugin.stages.DeclarativeContent`
+
+    Returns:
+        The content_unit_saver stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content_list = []
+    shutdown = False
+    while True:
+        try:
+            declarative_content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content_list and not shutdown:
+                declarative_content = await in_q.get()
+                declarative_content_list.append(declarative_content)
+                continue
+        else:
+            declarative_content_list.append(declarative_content)
+            continue
+
+        content_artifact_bulk = []
+        remote_artifact_bulk = []
+        remote_artifact_map = {}
+
+        with transaction.atomic():
+            for declarative_content in declarative_content_list:
+                if declarative_content is None:
+                    shutdown = True
+                    continue
+                if declarative_content.content.pk is None:
+                        declarative_content.content.save()
```

Comment:
So, `Content` is not saved in bulk?  Seems like an opportunity for a big performance improvement here.

Reply:
Yeah `Content` currently is not compatible with bulk save :( It throws an error because bulk_save() and multi-table inheritance are incompatible. This is actually the main thing I'm working on now: https://pulp.plan.io/issues/3812